### ISSUE TITLE
What is the actual API?

### ISSUE THREAD

**Issue by markshannon** [hpyproject/hpy#323]
If I were to start writing a new module in C, I wouldn't know where to start.
The actual header file seems to be quite deep in the source tree at https://github.com/hpyproject/hpy/blob/master/hpy/devel/include/hpy.h which includes a lot of other files, so I'm already a bit lost.

The docs at https://docs.hpyproject.org/en/latest/# describe the high-level concepts, porting and various other bits and pieces, but still no actual API.

The CPython API, for all its flaws, is more navigable thanks to the online documentation at https://docs.python.org/3/c-api/index.html.

HPy may provide a better API than CPython's, but without documentation, it is a challenge to use.
I wouldn't recommend duplicating all the API in parallel docs, like CPython does, but adding some parseable comments to the source and using a tool like doxygen would produce something usable.

**Comment by mattip** [comment]
#396 improved documentation. Please take a look, especially at the [api reference](https://docs.hpyproject.org/en/latest/api-reference/index.html) section. Of course there is more work needed, especially to fill out the [raw list](https://docs.hpyproject.org/en/latest/api-reference/function-index.html) of API functions, additional suggestions are welcome.

**Comment by markshannon** [comment]
So where is the API? I still can't find it without digging around.
It still isn't discoverable from the [README](https://github.com/hpyproject/hpy#readme), or at https://docs.hpyproject.org/en/latest/

**Comment by hodgestar** [comment]
@markshannon https://docs.hpyproject.org/en/latest/api-reference/

**Comment by markshannon** [comment]
Let me rephrase.
~I~ A developer new to HPy can't find find the API  without digging around.
https://docs.hpyproject.org/en/latest/api-reference/ is not easily found from https://docs.hpyproject.org/en/latest/index.html and not at all findable from the GitHub README.

Even https://docs.hpyproject.org/en/latest/api-reference/ says
"The Core API as defined in file public_api.h" but `public_api.h` is not a link.

HPy is "a better API for Python", so it seems odd that the actual API comes after discussions of design philosophy and links to the IRC channel.

**Comment by mattip** [comment]
The top-level [documentation link](https://docs.hpyproject.org/en/latest/index.html) has links called [HPy API introduction](https://docs.hpyproject.org/en/latest/api.html), and [API Reference](https://docs.hpyproject.org/en/latest/api-reference/index.html). Should those be moved up higher on the page? The function declarations are not very useful on their own, one needs a bit of background about the wrapping macros, a context, and managing handles. A [list of the functions](https://docs.hpyproject.org/en/latest/api-reference/function-index.html) exists and indeed future work will add more context to that list. Would you like to see that list have more prominence?

**Comment by mattip** [comment]
The README currently is more a "motivation for HPy" page than an overview of the project. Perhaps it is time to refactor it.

**Comment by mattip** [comment]
> A developer new to HPy ...

I think we view the path a new user would traverse would begin with samples of code using it, and to read a bit about the ideals behind it before getting to a list of API functions. Much like the CPython [extending](https://docs.python.org/3/extending/index.html) documentation.

**Comment by markshannon** [comment]
It's your project, so ultimately its up to you.
Updating the README to give more prominence to the the API link sounds like a good idea.
